versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapter.java

```diff
@@ -0,0 +1,401 @@
+      if (ErrorCategory.DUPLICATE_KEY.equals(category)) {
+        ReferenceConflictException ex = DatabaseAdapterUtil.hashCollisionDetected();
+        ex.initCause(writeException);
+        throw ex;
+      }
+
+      throw writeException;
+    }
+
+    if (!result.wasAcknowledged()) {
+      throw new IllegalStateException("Unacknowledged write to " + collection.getNamespace());
+    }
+  }
+
+  private static void insert(MongoCollection<Document> collection, List<Document> docs)
+      throws ReferenceConflictException {
+    if (docs.isEmpty()) {
+      return; // Mongo does not accept empty args to insertMany()
+    }
+
+    InsertManyResult result;
+    try {
+      result = collection.insertMany(docs);
+    } catch (MongoWriteException writeException) {
+      ErrorCategory category = writeException.getError().getCategory();
+      if (ErrorCategory.DUPLICATE_KEY.equals(category)) {
+        ReferenceConflictException ex = DatabaseAdapterUtil.hashCollisionDetected();
+        ex.initCause(writeException);
+        throw ex;
+      }
+
+      throw writeException;
+    }
+
+    if (!result.wasAcknowledged()) {
+      throw new IllegalStateException("Unacknowledged write to " + collection.getNamespace());
+    }
+  }
+
+  private void delete(MongoCollection<Document> collection, Set<Hash> ids) {
+    DeleteResult result = collection.deleteMany(Filters.in("_id", toIds(ids)));
+
+    if (!result.wasAcknowledged()) {
+      throw new IllegalStateException("Unacknowledged write to " + collection.getNamespace());
+    }
+  }
+
+  private byte[] loadById(MongoCollection<Document> collection, String id) {
+    Document doc = collection.find(Filters.eq(id)).first();
+    if (doc == null) {
+      return null;
+    }
+
+    Binary data = doc.get("data", Binary.class);
+    if (data == null) {
+      return null;
+    }
+
+    return data.getData();
+  }
+
+  private <T> T loadById(MongoCollection<Document> collection, Hash id, Parser<T> parser) {
+    return loadById(collection, toId(id), parser);
+  }
+
+  private <T> T loadById(MongoCollection<Document> collection, String id, Parser<T> parser) {
+    byte[] data = loadById(collection, id);
+    if (data == null) {
+      return null;
+    }
+
+    try {
+      return parser.parse(data);
+    } catch (InvalidProtocolBufferException e) {
+      throw new IllegalStateException(e);
+    }
+  }
+
+  /**
+   * Fetches a collection of documents by hash and returns them in the order of hashes requested.
+   *
+   * <p>Note: unknown hashes will correspond to {@code null} elements in the result list.
+   */
+  private <T> List<T> fetchMappedPage(
+      MongoCollection<Document> collection, List<Hash> hashes, Function<Document, T> mapper) {
+    List<String> ids = hashes.stream().map(this::toId).collect(Collectors.toList());
+    FindIterable<Document> docs = collection.find(Filters.in("_id", ids)).limit(hashes.size());
+
+    HashMap<Hash, Document> loaded = new HashMap<>(hashes.size());
```

Comment:
The initial-capacity is wrong here. You need `hashes.size() * 4 / 3 + 1` here.

Reply:
Good point. Fixed in e6b4f6eebea447420690ab610184702adf28f57c